src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -321,9 +322,14 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
                                 case ConversionKind.IntegerToPointer:
                                     current = conv.Operand;
                                     break;
+
                                 case ConversionKind.ExplicitUserDefined:
                                 case ConversionKind.ImplicitUserDefined:
+                                // expression trees rewrite this later.
+                                // it is a kind of user defined conversions on IntPtr and in some cases can fail
+                                case ConversionKind.IntPtr:
```

Comment:
If we are not in an expression tree, it IntPtr will cause an assert in codegen. And that one is more reliable since it is not conditional on optimizations being performed. 
Also, this method does not have enough context to know we are in an expression tree and I do not want to change that just to get an assert earlier, sometimes.